File: src/main/python/com/hazza/leetcode_by_python/SegmentTwo/No61_Rotate_List.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/6/11 19:48
# https://leetcode.com/problems/rotate-list/
from src.main.python.com.hazza.LeetCodeByPython.data_structure.ListNode import ListNode, makeListNode, printListNode


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or k < 1:
            return head
        pre = ListNode(0)
        pre.next = head
        cur, tail = pre, pre
        cnt = 0

        while tail.next:
            tail = tail.next
            cnt += 1

        k = cnt - k % cnt
        cur = head
        for _ in range(1, k):
            cur = cur.next
        tail.next = pre.next
        pre.next = cur.next
        cur.next = None

        return pre.next


if __name__ == '__main__':
    solution = Solution()

    node = makeListNode([1, 2, 3, 4, 5])
    res = solution.rotateRight(node, 2)
    printListNode(res)

    node = makeListNode([0, 1, 2])
    res = solution.rotateRight(node, 4)
    printListNode(res)

    node = makeListNode([1])
    res = solution.rotateRight(node, 1)
    printListNode(res)

File: src/main/java/com/hazza/leetcode/SegmentTwo/No61RotateList.java
package com.hazza.leetcode.SegmentTwo;

import com.hazza.leetcode.datastructure.ListNode;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/rotate-list/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-17
 * Time: 10:35 AM
 */
public class No61RotateList {
    public ListNode rotateRight(ListNode head, int k) {
        if (k < 1 || head == null) return head;
        ListNode temp = new ListNode(0);
        temp.next = head;
        ListNode cur = temp, tail = temp;
        int count = 0;
        while (tail.next != null) {
            tail = tail.next;
            ++count;
        }
        for (int i = count - k % count; i > 0; i--)
            cur = cur.next;
        tail.next = temp.next;
        temp.next = cur.next;
        cur.next = null;

        return temp.next;
    }
}
